[SPARK-16278][SPARK-16279][SQL] Implement map_keys/map_values SQL functions

dongjoon-hyun opened the pull request:
## What changes were proposed in this pull request?

This PR adds `map_keys` and `map_values` SQL functions in order to remove Hive fallback.
## How was this patch tested?

Pass the Jenkins tests including new testcases.

dongjoon-hyun committed:
[SPARK-16278][SPARK-16279][SQL] Implement map_keys/map_values SQL functions

dongjoon-hyun commented:
cc @rxin and @cloud-fan .

SparkQA commented:
**[Test build #61454 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61454/consoleFull)** for PR 13967 at commit [`2270b39`](https://github.com/apache/spark/commit/2270b395f2838e0076dc8ef56e5e32ee44f0f820).

SparkQA commented:
**[Test build #61454 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61454/consoleFull)** for PR 13967 at commit [`2270b39`](https://github.com/apache/spark/commit/2270b395f2838e0076dc8ef56e5e32ee44f0f820).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class MapKeys(child: Expression)`
  * `case class MapValues(child: Expression)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61454/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -44,6 +44,56 @@ case class Size(child: Expression) extends UnaryExpression with ExpectsInputType
+  override def foldable: Boolean = child.foldable
+
+  override def nullSafeEval(map: Any): Any = {
+    map.asInstanceOf[MapData].valueArray().copy()
```
What if we don't call `copy` here? It looks reasonable to copy, I'm just curious :)

dongjoon-hyun replied:
I actually didn't found the corresponding case. Is it safe?

cloud-fan replied:
A similar one is `GetStructField`, it can get an `array` column from a row without copying it. Maybe it's safe to not copy here too. cc @davies

dongjoon-hyun replied:
Thank you for confirming. That's nice.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -44,6 +44,56 @@ case class Size(child: Expression) extends UnaryExpression with ExpectsInputType
+ * Returns an unordered array containing the keys of the map.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(map) - Returns an unordered array containing the keys of the map.")
```
example would be great.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -44,6 +44,56 @@ case class Size(child: Expression) extends UnaryExpression with ExpectsInputType
+ * Returns an unordered array containing the values of the map.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(map) - Returns an unordered array containing the values of the map.")
```
example would be great.

dongjoon-hyun committed:
Remove redundant copy() and add function extended description.

SparkQA commented:
**[Test build #61536 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61536/consoleFull)** for PR 13967 at commit [`6953b4e`](https://github.com/apache/spark/commit/6953b4e6f1fdbcabdfebc6dfc8ad4035320cfca5).

dongjoon-hyun commented:
Thank you for review, @rxin and @cloud-fan .
According to the comments, I removed the redundant `copy` operations and added function detail description examples.

SparkQA commented:
**[Test build #61536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61536/consoleFull)** for PR 13967 at commit [`6953b4e`](https://github.com/apache/spark/commit/6953b4e6f1fdbcabdfebc6dfc8ad4035320cfca5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61536/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -44,6 +44,58 @@ case class Size(child: Expression) extends UnaryExpression with ExpectsInputType
+
+  override def dataType: DataType = ArrayType(child.dataType.asInstanceOf[MapType].keyType)
+
+  override def foldable: Boolean = child.foldable
```
this is the default of `UnaryExpression`

dongjoon-hyun replied:
Yep.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionFunctionsSuite.scala:
```diff
@@ -44,6 +44,16 @@ class CollectionFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
 
+  test("MapKeys/MapValues") {
+    val m0 = Literal.create(Map("a" -> "1", "b" -> "2"), MapType(StringType, StringType))
+    val m1 = Literal.create(Map[String, String](), MapType(StringType, StringType))
```
also check null case please

dongjoon-hyun replied:
Sure.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -2753,6 +2753,22 @@ object functions {
+   * @group collection_funcs
+   * @since 2.1.0
+   */
+  def map_keys(e: Column): Column = withExpr { MapKeys(e.expr) }
```
Let's not register these functions here for now, we will think about it later.

dongjoon-hyun replied:
Oh, I think that policy is about `xpath-` related stuff. Isn't it?

cloud-fan replied:
it's about all hive-fallback functions

dongjoon-hyun replied:
I see. Thank you for confirming!

dongjoon-hyun replied:
I was confused due to `posexplode` before. Now, it's clear. I'll remove soon.

dongjoon-hyun committed:
Add a null map testcase and remove redundant implementation.

dongjoon-hyun commented:
I added a `null` map testcase and remove redundant implementation.

For the removal from `sql/functions.scala`, there is no problem to remove that. But, could you check that again?

SparkQA commented:
**[Test build #61553 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61553/consoleFull)** for PR 13967 at commit [`3881a98`](https://github.com/apache/spark/commit/3881a98eaf6680d6493fabab4dde405abe987d0a).

SparkQA commented:
**[Test build #61553 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61553/consoleFull)** for PR 13967 at commit [`3881a98`](https://github.com/apache/spark/commit/3881a98eaf6680d6493fabab4dde405abe987d0a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61553/
Test PASSed.

dongjoon-hyun committed:
Remove from `sql.functions`.

SparkQA commented:
**[Test build #61584 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61584/consoleFull)** for PR 13967 at commit [`56d838e`](https://github.com/apache/spark/commit/56d838e20c1249f99c5175c744c072ad0eecdb92).

cloud-fan commented:
LGTM, pending jenkins

SparkQA commented:
**[Test build #61584 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61584/consoleFull)** for PR 13967 at commit [`56d838e`](https://github.com/apache/spark/commit/56d838e20c1249f99c5175c744c072ad0eecdb92).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61584/
Test PASSed.

dongjoon-hyun commented:
Hi, @cloud-fan .
Could you merge this PR? :)

dongjoon-hyun commented:
Rebased to the master.

SparkQA commented:
**[Test build #61661 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61661/consoleFull)** for PR 13967 at commit [`8db1e65`](https://github.com/apache/spark/commit/8db1e656f27aa1647fca7c86405959262c3365fd).

SparkQA commented:
**[Test build #61661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61661/consoleFull)** for PR 13967 at commit [`8db1e65`](https://github.com/apache/spark/commit/8db1e656f27aa1647fca7c86405959262c3365fd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61661/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

dongjoon-hyun commented:
Thank you, @cloud-fan and @rxin !
